dbms/src/Common/UInt128.h

```diff
@@ -25,19 +24,43 @@ struct UInt128
     UInt64 first;
     UInt64 second;
 
-    bool operator== (const UInt128 rhs) const { return first == rhs.first && second == rhs.second; }
-    bool operator!= (const UInt128 rhs) const { return first != rhs.first || second != rhs.second; }
-
-    bool operator== (const UInt64 rhs) const { return first == rhs && second == 0; }
-    bool operator!= (const UInt64 rhs) const { return first != rhs || second != 0; }
+    UInt128() = default;
+    explicit UInt128(const UInt64 rhs) : first(0), second(rhs) { }
+
+    bool operator== (const UInt128 & rhs) const { return second == rhs.second && first == rhs.first; }
+    bool operator!= (const UInt128 & rhs) const { return second != rhs.second || first != rhs.first; }
+    bool operator>= (const UInt128 & rhs) const { return (first == rhs.first) ? second >= rhs.second : first > rhs.first; }
+    bool operator>  (const UInt128 & rhs)  const { return (first == rhs.first) ? second > rhs.second : first > rhs.first; }
+    bool operator<= (const UInt128 & rhs) const { return (first == rhs.first) ? second <= rhs.second : first < rhs.first; }
+    bool operator<  (const UInt128 & rhs)  const { return (first == rhs.first) ? second < rhs.second : first < rhs.first; }
+
+    /** Type stored in the database will contains no more than 64 bits at the moment, don't need
+     *  to check the `second` element 
+     */
+    template<typename T> bool operator== (const T rhs) const { return first == 0 && static_cast<T>(second) == rhs; }
+    template<typename T> bool operator!= (const T rhs) const { return first != 0 || static_cast<T>(second) != rhs; }
+    template<typename T> bool operator>= (const T rhs) const { return first != 0 && static_cast<T>(second) >= rhs; }
+    template<typename T> bool operator>  (const T rhs)  const { return first != 0 && static_cast<T>(second) >  rhs; }
+    template<typename T> bool operator<= (const T rhs) const { return first == 0 && static_cast<T>(second) <= rhs; }
+    template<typename T> bool operator<  (const T rhs)  const { return first == 0 && static_cast<T>(second) >  rhs; }
+
+    template<typename T> explicit operator T() const { return static_cast<T>(second); }
+    operator UInt128() const { return *this; }
 
 #if !__clang__
 #pragma GCC diagnostic pop
 #endif
 
-    UInt128 & operator= (const UInt64 rhs) { first = rhs; second = 0; return *this; }
+    UInt128 & operator= (const UInt64 rhs) { first = 0; second = rhs; return *this; }
```

Comment:
So I just take a look that this method was only used in the file `Common/HashTable/HashTable.h` in the `ZeroTraits::set` method who initialize to zero so it should be safe. :)

(most of the code call directly the `first` and `second` attributes)